compactor/periodic.go

```diff
@@ -61,50 +61,94 @@ func newPeriodic(clock clockwork.Clock, h time.Duration, rg RevGetter, c Compact
 	return t
 }
 
-// periodDivisor divides Periodic.period in into checkCompactInterval duration
-const periodDivisor = 10
-
 // Run runs periodic compactor.
 func (t *Periodic) Run() {
-	interval := t.period / time.Duration(periodDivisor)
-	go func() {
-		initialWait := t.clock.Now()
-		for {
-			t.revs = append(t.revs, t.rg.Rev())
-			select {
-			case <-t.ctx.Done():
-				return
-			case <-t.clock.After(interval):
-				t.mu.Lock()
-				p := t.paused
-				t.mu.Unlock()
-				if p {
-					continue
-				}
-			}
+	go t.run()
+}
 
-			// wait up to initial given period
-			if t.clock.Now().Sub(initialWait) < t.period {
-				continue
-			}
+// periodically fetches revisions and ensures that
+// first element is always up-to-date for retention window
+func (t *Periodic) run() {
+	initialWait := t.clock.Now()
+	fetchInterval := t.getInterval()
+	retryInterval := t.getRetryInterval()
+
+	// e.g. period 9h with compaction period 1h, then retain up-to 9 revs
+	// e.g. period 12h with compaction period 1h, then retain up-to 12 revs
+	// e.g. period 20m with compaction period 20m, then retain up-to 1 rev
+	retentions := int(t.period / fetchInterval)
+	for {
+		t.revs = append(t.revs, t.rg.Rev())
+		if len(t.revs) > retentions {
+			t.revs = t.revs[1:]
+		}
 
-			rev, remaining := t.getRev()
-			if rev < 0 {
+		select {
+		case <-t.ctx.Done():
+			return
+		case <-t.clock.After(fetchInterval):
+			t.mu.RLock()
+			p := t.paused
+			t.mu.RUnlock()
+			if p {
 				continue
 			}
+		}
 
+		// no compaction until initial wait period
+		if t.clock.Now().Sub(initialWait) < t.period {
+			continue
+		}
+		rev := t.revs[0]
+
+		for i := 0; i < retryDivisor; i++ {
 			plog.Noticef("Starting auto-compaction at revision %d (retention: %v)", rev, t.period)
 			_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
 			if err == nil || err == mvcc.ErrCompacted {
-				// move to next sliding window
-				t.revs = remaining
+				// compactor succeeds at revs[0], move sliding window
+				t.revs = t.revs[1:]
 				plog.Noticef("Finished auto-compaction at revision %d", rev)
-			} else {
-				plog.Noticef("Failed auto-compaction at revision %d (%v)", rev, err)
-				plog.Noticef("Retry after %v", interval)
+				break
+			}
+
+			// compactor fails at revs[0]:
+			//  1. retry revs[0], so long as revs[0] is up-to-date
+			//  2. retry revs[1], when revs[0] becomes stale
+			plog.Noticef("Failed auto-compaction at revision %d (%v)", rev, err)
+			plog.Noticef("Retry after %v", retryInterval)
+			select {
+			case <-t.ctx.Done():
+				return
+			case <-t.clock.After(retryInterval):
+				t.mu.RLock()
+				p := t.paused
+				t.mu.RUnlock()
+				if p {
+					break
```

Comment:
this `break` won't stop the loop, because it's captured by the `select`.

Reply:
you can move the paused to the top of the loop and maybe merge another one at line 90 as well.

Reply:
oh right. good catch. fixed.